keps/sig-storage/20200127-recover-from-resize.md

```diff
@@ -0,0 +1,153 @@
+
+To solve aformentioned problem - we propose that, a new field will be added to PVC, called `pvc.Spec.AllocatedResources`. This field is only allowed to increase and will be set by the api-server to value in `pvc.Spec.Resources` as long as `pvc.Spec.Resources > pvc.Spec.AllocatedResources`.  Quota controller code will be updated to use `max(pvc.Spec.Resources, pvc.Spec.AllocatedResources)` when calculating usage of the PVC under questions. This does mean that - if user expanded PVC to a size that it failed to expand to and user retries expansion with lower value, the used quota will still use value reported by failed expansion request.
+
+In practice above limitation should be fine. We will add events to PV and PVC when we detect that user has reduced requested size. Admin can recover from this state either by manually increasing quota or by manually rebuilding the PVC object.
+
+#### User flow stories
+
+###### Case 1 (controller+node expandable):
+
+- User increases 10Gi PVC to 100Gi
+- Api Server sets `pvc.spec.allocatedResources` to 100Gi.
+- Quota controller uses `max(pvc.Spec.AllocatedResources, pvc.Spec.Resources)` and adds `90Gi` to used quota.
+- Expansion to 100Gi fails and hence `pv.Spec.Capacity` and `pvc.Status.Capacity `stays at 10Gi.
+- User requests size to 20Gi.
+- Quota controller sees no change in storage usage by the PVC because `pvc.Spec.AllocatedResources` is `100Gi`.
+- Expansion succeeds and `pvc.Status.Capacity` and `pv.Spec.Capacity` report new size as `20Gi`.
+- `pvc.Spec.AllocatedResources` however keeps reporting `100Gi`.
+
+
+###### Case 2 (controller+node expandable):
+
+- User increases 10Gi PVC to 100Gi
+- Api Server sets `pvc.spec.allocatedResources` to 100Gi.
+- Quota controller uses `max(pvc.Spec.AllocatedResources, pvc.Spec.Resources)` and adds `90Gi` to used quota.
+- Expansion to 100Gi fails and hence `pv.Spec.Capacity` and `pvc.Status.Capacity `stays at 10Gi.
+- User requests size to 20Gi.
+- Quota controller sees no change in storage usage by the PVC because `pvc.Spec.AllocatedResources` is `100Gi`.
+- Expansion succeeds and `pvc.Status.Capacity` and `pv.Spec.Capacity` report new size as `20Gi`.
+- `pvc.Spec.AllocatedResources` however keeps reporting `100Gi`.
+- User expands PVC to 120Gi
+- Api Server sets `pvc.spec.allocatedResources` to 120Gi.
+- Quota controller uses `max(pvc.Spec.AllocatedResources, pvc.Spec.Resources)` and adds `20Gi` to used quota.
+- Expansion to 120Gi fails and hence `pv.Spec.Capacity` and `pvc.Status.Capacity `stays at 20Gi.
+
+### Risks and Mitigations
+
+- One risk as mentioned above is, if expansion failed and user retried expansion(successfully) with smaller value, the quota code will keep reporting higher value. In practice though - this should be acceptable since such expansion failures should be rare and admin can unblock the user by increasing the quota or rebuilding PVC if needed. We will emit events on PV and PVC to alerts admins and users.
```

Comment:
Is there any follow-up we can do to automatically reconcile quota or fix this weird state or will it just persist in the cluster "forever" until someone realizes something is wrong and has to go in and manually fix it.

Reply:
We will document about how to fix the quota but I do not think there will be a mechanism possible to reconcile the quota automatically. Added something similar in alternatives considered section.